polkadot/node/core/pvf/src/execute/queue.rs

```diff
@@ -638,3 +653,299 @@ pub fn start(
 	.run();
 	(to_queue_tx, from_queue_rx, run)
 }
+
+struct Unscheduled {
+	unscheduled: HashMap<PvfExecPriority, VecDeque<ExecuteJob>>,
+	counter: HashMap<PvfExecPriority, usize>,
+}
+
+impl Unscheduled {
+	/// We keep track of every scheduled job in the `counter`, but reset it if the total number of
+	/// counted jobs reaches the threshold. This number is set as the maximum amount of jobs per
+	/// relay chain block possible with 4 CPU cores and 2 seconds of execution time. Under normal
+	/// conditions, the maximum expected queue size is at least vrf_module_samples(6) + 1 for
+	/// backing a parachain candidate. A buffer is added to cover situations where more work
+	/// arrives in the queue.
+	const SCHEDULING_WINDOW_SIZE: usize = 12;
+
+	/// A threshold in percentages indicates how much time a current priority can "steal" from lower
+	/// priorities. Given the `SCHEDULING_WINDOW_SIZE` is 12 and all job priorities are present:
+	/// - Disputes consume 70% or 8 jobs in a row.
+	/// - The remaining 30% of original 100% is allocated for approval and all backing jobs.
+	/// - 80% or 3 jobs of the remaining goes to approvals.
+	/// - The remaining 6% of original 100% is allocated for all backing jobs.
+	/// - 100% or 1 job of the remaining goes to backing system parachains.
+	/// - Nothing is left for backing.
+	/// - The counter is restarted and the distribution starts from the beginning.
+	///
+	/// This system might seem complex, but we operate with the remaining percentages because:
+	/// - Not all job types are present in each block. If we used parts of the original 100%,
+	///   approvals could not exceed 24%, even if there are no disputes.
+	/// - We cannot fully prioritize backing system parachains over backing other parachains based
+	///   on the distribution of the original 100%.
```

Comment:
Yeah, we might've some gradiations among the different system parachains, so do whatever you think is best for now.  AssetHub could maybe halt without killing anything, although this depends upon how its used.  Among the chains that must run "enough" during an epoch:
- Sassafras tickets collection, if we ever moved it onto a parachain (maybe never).
- Validator elections should maintain a fresh validator set, even if we've lost many for some reason.
- If validator elections runs then we must run bridgehub enough, and any future DKG chains.
- Some governance chain without smart contracts (collectives?).

Reply:
Yeah this is just a first step. I agree it needs to be e reconsidered once we start out to move governance and staking out of the relay chain.

However, I think that a better algorithm would actually consider to target specific block times for the system parachains in times of overload.

Reply:
Afaik, critical system paracahins should typically not require numerous blocks, specific block times, etc, but some require like one block from each validator during the epoch, including enough slack so spam cannot censor them.

This assumes the critical system paracahins cannot be tricked into wasting their own blocks, which likely becomes a risk only if the d-day governance supports smart contracts.